### PR TITLE
cleanup: resolve compiler warnings

### PR DESCRIPTION
**Pull request by adamemerson**
There are 2×10¹⁴ metres of DNA in all the cells of my body. If every run of four base-pairs encoded the word 'hate'…okay that's a bit overboard. Sure, I don't *like* warnings, but there are a lot of things that deserve to have hate lavished upon them more than compiler diagnostics do.

### CODE REVIEW

File: src/librbd/ExclusiveLock.h [hunk 1]
```diff
@@ -17,7 +17,7 @@ class ExclusiveLock : public ManagedLock<ImageCtxT> {
   }
 
   ExclusiveLock(ImageCtxT &image_ctx);
-
+  ~ExclusiveLock() = default;
```
Comment (tchaikov) [inline src/librbd/ExclusiveLock.h]: this is not necessary, right? as `LeaderLock` will have a virtual default dtor anyway. and so does its child classes if any. we can even mark `LeaderLock` `final` if we want to be more explicit.

File: src/tools/rbd_mirror/LeaderWatcher.h [hunk 2]
```diff
@@ -98,6 +98,8 @@ class LeaderWatcher : protected librbd::Watcher {
         watcher(watcher) {
     }
 
+    virtual ~LeaderLock() = default;
```
Comment (tchaikov) [inline src/tools/rbd_mirror/LeaderWatcher.h]: ditto.

File: src/librbd/ExclusiveLock.h (hunk 1 above)
Comment (tchaikov) [inline src/librbd/ExclusiveLock.h]: @dillaman do you mind taking a look?

**Comment by tchaikov** [comment]
modulo the nit, lgtm.

File: src/tools/rbd_mirror/LeaderWatcher.h (hunk 2 above)
Comment (adamemerson) [inline src/tools/rbd_mirror/LeaderWatcher.h]: @tchaikov The default destructor is never generated as virtual. If you want a virtual one you have to specify it. (And the compiler was complaining about it not being virtual.)

Comment (tchaikov) [inline src/tools/rbd_mirror/LeaderWatcher.h]: @adamemerson `LeaderLock` derives from `librbd::ManagedLock<ImageCtxT>`, and the latter does have a virtual dtor. so `LeaderLock` will have a virtual dtor even if we don't put
```c++
virtual ~LeaderLock() = default;
```
here, am i right?

File: src/librbd/ExclusiveLock.h (hunk 1 above)
Comment (dillaman) [inline src/librbd/ExclusiveLock.h]: Nit: base class [1] has a virtual destructor, so by definition this already has a virtual destructor

[1] https://github.com/adamemerson/ceph/blob/1dc449ac5d8ed087af677b61d6da94113bbdf7e4/src/librbd/ManagedLock.h#L47

File: src/tools/rbd_mirror/LeaderWatcher.h (hunk 2 above)
Comment (dillaman) [inline src/tools/rbd_mirror/LeaderWatcher.h]: Nit: base class [1] has a virtual destructor, so by definition this already has a virtual destructor

[1] https://github.com/adamemerson/ceph/blob/1dc449ac5d8ed087af677b61d6da94113bbdf7e4/src/librbd/ManagedLock.h#L47

Comment (dillaman) [inline src/tools/rbd_mirror/LeaderWatcher.h]: Think you wanted to put that in the mocked back class [1]

[1] https://github.com/adamemerson/ceph/blob/1dc449ac5d8ed087af677b61d6da94113bbdf7e4/src/test/rbd_mirror/test_mock_LeaderWatcher.cc#L31

**Comment by adamemerson** [comment]
@dillaman I have changed it to add virtual destructors to exactly and only the base classes.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-02-07_15:35:38-rados-wip-kefu-testing---basic-mira/

**Pull request merged**